src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentsState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentsState = {};
+        break;
+      case 'addProperties':
+        currentsState = { ...currentsState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentsState = { ...currentsState };
+
+        action.keysToRemove.forEach((key) => {
+          delete currentsState[key];
+        });
+        break;
+    }
+    stateHistory.push(currentsState);
```

Comment:
It is better to push a copy of the object to the `stateHistory` array to prevent all elements in the array from referencing the same object. This can be done by pushing `{ ...currentState }` instead of `currentState`. [CODE STYLE]